### DISCUSSION TITLE
Debugging tool with breakpoints

### DISCUSSION THREAD

**Discussion by sk8board** [micropython/micropython#16037] [Core Development] [answered]
Is there a debugging tool with breakpoints available for MicroPython? 

If not, is there a plan to develop a debugging tool with breakpoints?

**Comment by andrewleech** [comment]
I've been gradually working towards full debugger support in micropython, however it's a stop/start process when I get time: https://github.com/micropython/micropython-lib/pull/499

I've seen a few people use `sys.settrace` to build simple debugging tools before but can't find them right now, I haven't really used any myself.

**Reply by Omid888** [reply]
Having a debugger could benefit everyone. Thank you.

**Comment by sk8board** [comment]
Thank you for your reply. Is there a tutorial or video demo of how this debugger functions?

**Comment by andrewleech** [comment, marked as answer]
I don't have any docs or similar yet, there's still some groundwork to finish before it's usable. 

This is intended to provide `pdb` which is the debugging module from cpython. You can google how to use it manually, but it's also the engine used by pretty much all standard python debugging tools so I'm hoping to work towards getting vscode debugging working on top of this.

**Comment by gus-ghielec** [comment]
This is solved now, if anyone finding this issue wants it. Not PDB, a debug channel in the firmware plus a VS Code extension speaking DAP.

Breakpoints (conditional, and inside .mpy), step over/into/out, call stack, locals and globals, watch expressions, and print() output, over the existing USB connection. No JTAG or probe, just F5, F10, F11.

Generic firmware builds for RP2040, RP2350, ESP32-S2 and ESP32-S3, and the extension can flash them for you. Source for the firmware integration is in a separate repo if you want to build for your own board.

Extension: https://marketplace.visualstudio.com/items?itemName=ghielectronics.micropython-debugger
Source: https://github.com/ghi-electronics/micropython-vsc-extension (MIT)

The protocol and firmware side are MIT and work today. We're open to upstreaming it if that's useful to the project — happy to adapt the design to whatever maintainers would want to maintain long-term.
